File: algorithms/3_hard/M/01383_maximum_performance_of_a_team/heap.cpp
typedef ptrdiff_t isize;

constexpr int64_t mod = 1'000'000'007;

class Solution {
public:
    static int maxPerformance(const int n, const vector<int>& speed, const vector<int>& efficiency, const int k_i32) {
        assert(k_i32 > 0);
        const isize k = k_i32;
        const isize len = std::ssize(speed);
        assert(len == std::ssize(efficiency));

        auto combo = vector<pair<int, int>>(len);

        for (isize i = 0; i < len; i += 1) {
            combo[i] = {efficiency[i], speed[i]};
        }

        ranges::sort(combo, [](const auto& a, const auto& b) {
            return a.first > b.first;
        });

        auto heap = priority_queue<int, vector<int>, greater<>>();
        int64_t max_res = 0;
        int64_t sum_speed = 0;

        for (const auto [e, s] : combo) {
            sum_speed += s;
            heap.push(s);
            if (std::ssize(heap) > k) {
                const int s_to_remove = heap.top();
                heap.pop();
                sum_speed -= s_to_remove;
            }

            const int64_t curr = sum_speed * e;
            max_res = std::max(max_res, curr);
        }

        return static_cast<int>(max_res % mod);
    }
};

File: algorithms/3_hard/M/01383_maximum_performance_of_a_team/heap.rs
use std::collections::BinaryHeap;
use std::cmp::Reverse;

impl Solution {
    const MOD_TO: i64 = 1_000_000_007;

    pub fn max_performance(n: i32, speed: Vec<i32>, efficiency: Vec<i32>, k: i32) -> i32 {
        assert!(k > 0);
        let k = k as usize;
        let mut combo = efficiency.iter().zip(speed.iter()).map(|(a, b)| (*a, *b)).collect::<Vec<_>>();
        combo.sort_unstable();

        let mut heap: BinaryHeap<Reverse<i32>> = BinaryHeap::with_capacity(k + 1);
        let mut max_res = 0_i64;
        let mut sum_speed = 0_i64;

        for &(e, s) in combo.iter().rev() {
            sum_speed += s as i64;
            heap.push(Reverse(s));
            if heap.len() > k {
                let Reverse(s_to_remove) = heap.pop().unwrap();
                sum_speed -= s_to_remove as i64;
            }

            let curr = sum_speed * (e as i64);
            max_res = max_res.max(curr);
        }

        (max_res % Self::MOD_TO).try_into().unwrap()
    }
}
